src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,433 @@
+/*-------------------------------------------------------------------------
+ *
+ * distributed_transaction_management.c
+ *
+ *  Infrastructure for distributed transaction management and distributed deadlock
+ *  detection.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "miscadmin.h"
+
+#include "funcapi.h"
+#include "access/htup_details.h"
+#include "catalog/pg_type.h"
+#include "datatype/timestamp.h"
+#include "distributed/distributed_transaction_management.h"
+#include "distributed/metadata_cache.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "storage/proc.h"
+#include "storage/spin.h"
+#include "storage/s_lock.h"
+#include "utils/timestamp.h"
+
+
+/*
+ * Each backend's active distributed transaction data reside in the
+ * shared memory on the DistributedTransactionShmemData.
+ */
+typedef struct DistributedTransactionShmemData
+{
+	int trancheId;
+#if (PG_VERSION_NUM >= 100000)
+	NamedLWLockTranche namedLockTranche;
+#else
+	LWLockTranche lockTranche;
+#endif
+	LWLock lock;
+
+	pg_atomic_uint64 nextTransactionId;
+
+	DistributedTransactionBackendData sessions[FLEXIBLE_ARRAY_MEMBER];
+} DistributedTransactionShmemData;
+
+
+static shmem_startup_hook_type prev_shmem_startup_hook = NULL;
+static DistributedTransactionShmemData *distributedTransactionShmemData = NULL;
+static DistributedTransactionBackendData *DistributedTransactionBackend = NULL;
+slock_t BackendDistributedTransactionMutex = '\0';
+
+
+static void DistributedTransactionManagementShmemInit(void);
+static Datum GenerateDistributedTransactionIdTuple(Oid databaseId, uint64
+												   initiatorNodeIdentifier, uint64
+												   transactionId, TimestampTz timestamp);
+static TupleDesc GenerateDistributedTransactionTupleDesc(void);
+static size_t DistributedTransactionManagementShmemSize(void);
+static uint64 GetNextLocalTransactionIdFromShmem(void);
+
+
+PG_FUNCTION_INFO_V1(assign_distributed_transaction_id);
+PG_FUNCTION_INFO_V1(get_distributed_transaction_id);
+
+
+/*
+ * assign_distributed_transaction_id updates the shared memory allocated for this backend
+ * and sets initiatorNodeIdentifier, transactionId, timestamp fields with the given
+ * inputs. Also, the function sets the database id via MyDatabaseId variable provided
+ * by "misadmmin.h".
+ */
+Datum
+assign_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	/* TODO: Is assert OK? */
```

Comment:
Should be, yes. Perhaps make it an ereport(ERROR) instead, just out of paranoia?